HW2/P2/parallel_vector.pyx

```diff
@@ -54,39 +55,85 @@ cpdef move_data_serial(np.int32_t[:] counts,
                        np.int32_t[:] dest,
                        int repeat):
-   cdef:
-       int idx, r
+	cdef:
+		int idx, r
 
-   assert src.size == dest.size, "Sizes of src and dest arrays must match"
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+	assert src.size == dest.size, "Sizes of src and dest arrays must match"
+	with nogil:
+		for r in range(repeat):
+			for idx in range(src.shape[0]):
+				if counts[src[idx]] > 0:
+					counts[dest[idx]] += 1
+					counts[src[idx]] -= 1
 
 
 cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] src,
                              np.int32_t[:] dest,
                              int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
+	cdef:
+		int idx, r
+		omp_lock_t *locks = get_N_locks(counts.shape[0])
 
    ##########
    # Your code here
    # Use parallel.prange() and a lock for each element of counts to parallelize
    # data movement.  Be sure to avoid deadlock, and double-locking.
    ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
 
-   free_N_locks(counts.shape[0], locks)
+	print('Number of locks for fine-grained: {}'.format(counts.shape[0])) 
+
+
+	for r in range(repeat):
+
+		for idx in prange(src.shape[0], nogil = True , num_threads = 4 ):
+
+
+			#--------------------------------------
+			# need to separate to avoid deadlocking
+			# ex: src[idx] = 5 and dest[idx] = 6 on thread 1
+			# while src[idx] = 6 and dest[idx] = 5 on thread 2
+			#--------------------------------------
+			if src[idx] < dest[idx]:
+           	# grab the lock 
+				acquire(&(locks[src[idx]])) 
+				acquire(&(locks[dest[idx]])) 
+           	# move data 
+				if counts[src[idx]] > 0:
+					counts[dest[idx]] += 1 
+					counts[src[idx]] -= 1    
+           	# release the lock 
+				release(&(locks[src[idx]]))
+				release(&(locks[dest[idx]]))    
+
+
+			elif src[idx] > dest[idx]:
+           	# grab the lock
+				acquire(&(locks[dest[idx]]))            
+				acquire(&(locks[src[idx]])) 
+           	# move data 
+				if counts[src[idx]] > 0:
+					counts[dest[idx]] += 1 
+					counts[src[idx]] -= 1    
+           	# release the lock 
+				release(&(locks[dest[idx]]))    
+				release(&(locks[src[idx]]))
+
+
+			#--------------------------------------
+            # here is to avoid double locking
+            #--------------------------------------
+			else:
+				acquire(&(locks[src[idx]])) 
+           	# move data
+				if counts[src[idx]] > 0:
+					counts[dest[idx]] += 1
+					counts[src[idx]] -= 1
+           	# release the lock
+				release(&(locks[src[idx]]))
+
```

Comment:
Rather than duplicating all this code, you could wrap just the locking code in your if/elseif/else construct, and have the actual data manipulation only once.